### ISSUE TITLE
Embedded markdown readme for NuGet packages

### ISSUE THREAD

**Issue by MSDN-WhiteKnight** [dotnet/runtime#59630]
NuGet introduced [embedded markdown Readme support](https://devblogs.microsoft.com/nuget/add-a-readme-to-your-nuget-package/) in May 2021. Now the updated gallery design displays this readme file on the package page if it's included. Looks like most runtime packages don't have readme embedded (neither markdown nor plaintext), and for them it fallbacks to description from nuspec. This leads to experience like [this](https://www.nuget.org/packages/System.Reflection.MetadataLoadContext/6.0.0-rc.1.21451.13):

![mlc](https://user-images.githubusercontent.com/35516665/134845772-53b25ab9-ffa5-4e0b-aba4-2f2e4870909f.png)

A lot of space is unused here. It could be filled with some more introductory information or simple usage example. And also listed "Commonly used types" could be links to their API docs. That's how package with markdown readme looks like:

![mysql_connector](https://user-images.githubusercontent.com/35516665/134846000-a92c80d3-5c50-455c-93d0-6a962bad9904.png)

Could we add markdown readme files for runtime packages?

---

Progress:

- [x] [microsoft.extensions.configuration](https://github.com/dotnet/runtime/pull/77733)
- [x] [microsoft.extensions.configuration.json](https://github.com/dotnet/runtime/pull/75876)
- [x] [system.configuration.configurationmanager](https://github.com/dotnet/runtime/pull/72772)
- [x] [system.text.json](https://github.com/dotnet/runtime/pull/69015)
- [x] [microsoft.extensions.configuration.binder](https://github.com/dotnet/runtime/pull/77649)
- [x] [microsoft.extensions.configuration.abstractions](https://github.com/dotnet/runtime/pull/77733)
- [x] [System.Runtime.Caching](https://github.com/dotnet/runtime/pull/90701)

For up-to-date checklist, see PR #91210

See also: 

- [Write a high-quality README for NuGet packages](https://devblogs.microsoft.com/nuget/write-a-high-quality-readme-for-nuget-packages/)
- [Package Readme guidelines](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/libraries-packaging.md#package-readme)
- [Example PR with new Readme template](https://github.com/dotnet/runtime/pull/90102)

**Comment by joperezr** [comment]
That is a good suggestion. Moving issue to infrastructure as the change to add markdown for package descriptions would likely go there. We can consider using third-party libraries that convert a (markdown) string into html. We could use one of those and then let individual packages onboard markdown description as they see fit.

**Comment by ViktorHofer** [comment]
Would this actually require any infrastructure support? I assume a library author just needs to set a property to point to the manually written markdown file.

**Comment by joperezr** [comment]
I'm not sure how this works, I assumed that in order for this to work then you should need to write the description in Html so that the Nuspec is able to parse it and show it in nuget.org. In any case, even if that is not the case and all we need to do is to change the description to markdown, I believe infrastructure has usually been the owners of package ddescription so I'd imagine that we should do an overall pass through all packages to make the conversion.

**Comment by svick** [comment]
@joperezr [The relevant documentation is here.](https://docs.microsoft.com/en-us/nuget/nuget-org/package-readme-on-nuget-org#supported-markdown-features) In short, the readme file has to be in Markdown, so no extra work for converting it is needed.

**Comment by ViktorHofer** [comment]
I'm wondering if it makes more sense to convert every library's existing PackageDescription property content into README.md markdown files or just keep it as is and let library authors decide if they would like to leverage the README.md support.

Pros for converting now:
- Easier to add content to what's already there

Cons:
- Adding a README.md file per project and shipping it via our packages even though most packages won't have any additional information that would warrant a README.md file.

**Comment by MSDN-WhiteKnight** [comment]
Better to convert individual libraries as needed, i think. This is mainly needed for out-of-band libraries that are explicitly installed via package manager. For libraries like System.Collections.Immutable that are inbox on all .NET Core versions and only installed via package manager for .NET Framework, it's less important. 

We also need to define convention where readme.md is placed in projects. Should it be `docs\README.md` as in NuGet default example, or just README.md near .csproj file?

**Comment by ViktorHofer** [comment]
> We also need to define convention where readme.md is placed in projects. Should it be docs\README.md as in NuGet default example, or just README.md near .csproj file?

No need for another folder. I would just put it next to the csproj file.

**Comment by joperezr** [comment]
I think that when it comes to infrastructure, one more pro to call out from @ViktorHofer list is consistency. Having all OOB packages use the same process to add the package description makes it easier to maintain and understand for the rest of the folks working on the repo, so my vote is either on moving all to README.md files, or not onboarding into markup, but not be in an in-between state.

**Comment by ViktorHofer** [comment]
There's nothing that is stopping library owners in dotnet/runtime from adding a README.md file and packaging it, given that we migrated all libraries from the old pkgproj infrastructure to the NuGet Pack task during NET6.

I don't think it's a good strategy to convert every library's existing PackageDescription content to README.md files as in many cases that description is quite short and doesn't offer any details on how to actually use the exposed APIs. Instead, it sounds like package consumers would want us to include "Getting Started" documentation per package, which doesn't exist today.

Based on that, I'm moving this to area-Meta as there's nothing that the infrastructure team needs to provide.

**Comment by teo-tsirpanis** [comment]
@MSDN-WhiteKnight can you remove `System.Net.Http` from the task list? This package is legacy.

**Comment by ViktorHofer** [comment]
Do we actually want to package the README.md files from the source tree? Those include details about what kind of source changes we accept, sometimes describe the dotnet/runtime developer innerloop; and such target a different audience than the one that consumes the package.

cc @ericstj

**Comment by MSDN-WhiteKnight** [comment]
Probably not all of them. Ones that i were adding also contain information for package users, for example: https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Binder/README.md

Microsoft.Extensions area owners preferred to consolidate library readme and package readme, so here it is one "merged" readme. But it is not the case for other libraries, for example, System.Text.Json now has two different readmes: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/README.md and https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/README.md

**Comment by ViktorHofer** [comment]
> Microsoft.Extensions area owners preferred to consolidate library readme and package readme, so here it is one "merged" readme.

Interesting. Are we sure that we want to leak the internal repo specific documentation to consumers on nuget.org? That doesn't sound right to me.

**Comment by jeffhandley** [comment]
> Do we actually want to package the README.md files from the source tree? Those include details about what kind of source changes we accept, sometimes describe the dotnet/runtime developer innerloop; and such target a different audience than the one that consumes the package.

Agreed; it was not intended for these recently-added READMEs to target library consumers. The audience for those is folks engaging within the repo source and the content is indeed different.

I could imagine having a `PACKAGE.md` at the root of the library folders where it's clear the purpose of that file is to be the markdown for the package.

**Comment by ViktorHofer** [comment]
Fixed by https://github.com/dotnet/runtime/pull/78888

**Comment by lyndaidaii** [comment]
Hi, I'm from the NuGet team. We are driving effort to increase the adoption and quality of NuGet package READMEs. 

Here is .NET library package candidates that are top installed in Visual Studio.

- System.Data.OleDb
- System.Data.Odbc
- System.Management
- System.DirectoryServices.AccountManagement
- System.DirectoryServices
- System.ServiceProcess.ServiceController
- System.Windows.Extensions
- System.CodeDom
- System.Diagnostics.EventLog
- System.Runtime.Caching
- System.IO.Ports
- Microsoft.Bcl.AsyncInterfaces
- System.Speech
- System.Text.Encodings.Web
- System.Security.Permissions
- System.Net.Http.Json
- System.Drawing.Common

To help you get started, we suggest the following steps to include a README in your packages listed below:
•	If you are new to NuGet README, learn more about [how to add a README to your NuGet Package](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdevblogs.microsoft.com%2Fnuget%2Fadd-a-readme-to-your-nuget-package%2F&data=05%7C01%7CLynn.Dai%40microsoft.com%7C104a7d575aa945c929e408db97e8d3cf%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638270798485909796%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=0OcAUfiee90eoUxIVEc0FLIh4juGuque56O3lBBdElw%3D&reserved=0).
•	Follow the [README template](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fraw.githubusercontent.com%2Flyndaidaii%2FNuGetReadMeTemplate%2Fmain%2FREADME.md&data=05%7C01%7CLynn.Dai%40microsoft.com%7C104a7d575aa945c929e408db97e8d3cf%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638270798485909796%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=XNUA17w6bfKoax%2FSIEAMRQMY%2Fk1E4wEwFzvSDK6M6BA%3D&reserved=0) and [blog post](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdevblogs.microsoft.com%2Fnuget%2Fwrite-a-high-quality-readme-for-nuget-packages%2F&data=05%7C01%7CLynn.Dai%40microsoft.com%7C104a7d575aa945c929e408db97e8d3cf%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638270798485909796%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=VT0JizwK%2B2mk5qz97r5W2NDNS9eRIplqEg%2BOsNab39Q%3D&reserved=0) for  best practices.
•	Check out a great [example](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fruntime%2Fpull%2F90102&data=05%7C01%7CLynn.Dai%40microsoft.com%7C104a7d575aa945c929e408db97e8d3cf%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638270798485909796%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=UbbSUPhoulewaR9S%2BspYeBS8tCYrTKBQKSPjYSxfGCI%3D&reserved=0) that created README for System.Text.Json that can be used for .NET libraries.
•	Reach out to us for review or help.

**Comment by danmoseley** [comment]
@ViktorHofer what does this track that's not in #91210

**Comment by ViktorHofer** [comment]
This issue existed before we started working on package readmes in https://github.com/dotnet/runtime/pull/91210. Let's close this issue in favor of the PR. After the PR is merged (this week), I'll open another issue for the remaining packages.
